storage/replica.go

```diff
@@ -507,32 +515,84 @@ func (r *Replica) SetLastVerificationTimestamp(timestamp proto.Timestamp) error
+		return
+	}
+	for i, union := range bArgs.Requests {
+		// If we did this for PushTxn as well, conflict resolution would break
+		// because a batch full of pushes of the same Txn for different intents
+		// will push the pushee in iteration, bumping up its priority.
+		// Unfortunately PushTxn is flagged as a write command (is it really?).
+		// Interim solution.
+		if args := union.GetValue().(proto.Request); args.Method() != proto.PushTxn {
+			args.Header().Timestamp.Forward(bArgs.Timestamp.Add(0, int32(i)))
+		}
+	}
+}
+
 // AddCmd adds a command for execution on this range. The command's
 // affected keys are verified to be contained within the range and the
 // range's leadership is confirmed. The command is then dispatched
 // either along the read-only execution path or the read-write Raft
 // command queue.
-func (r *Replica) AddCmd(ctx context.Context, args proto.Request) (proto.Response, error) {
+func (r *Replica) AddCmd(ctx context.Context, args proto.Request) (reply proto.Response, err error) {
+	// Wrap non-batch requests in a batch if possible.
+	bArgs, ok := args.(*proto.BatchRequest)
+	// TODO(tschottdorf): everything should go into a Batch. Keeping some commands
+	// individual creates a lot of trouble on the coordinator side for nothing.
+	// Instead, just enforce that some commands must be alone in their batch.
+	// Then they can just be unwrapped hereabouts (admin commands).
+	if !ok && proto.CanBatch(args) {
+		bArgs = &proto.BatchRequest{
+			RequestHeader: *args.Header(),
+		}
+		bArgs.Add(args)
+		// Unwrap reply via deferred function.
+		defer func() {
+			if err == nil {
+				bReply, ok := reply.(*proto.BatchResponse)
+				if !ok || len(bReply.Responses) != 1 {
+					panic("expected a BatchResponse with a single wrapped response")
+				}
+				reply = bReply.Responses[0].GetValue().(proto.Response)
+			} else {
+				reply = args.CreateReply()
+			}
+		}()
+	}
+	if bArgs != nil {
+		// Fiddle with the timestamps to make sure that writes can overlap
+		// within this batch.
+		// TODO(tschottdorf): provisional feature to get back to passing tests.
+		// Have to discuss how we go about it.
+		setBatchTimestamps(bArgs)
+	}
 	// TODO(tschottdorf) Some (internal) requests go here directly, so they
 	// won't be traced.
 	trace := tracer.FromCtx(ctx)
-	// Differentiate between admin, read-only and read-write.
-	var reply proto.Response
-	var err error
+	// Differentiate between admin, read-only and write.
 	if proto.IsAdmin(args) {
 		defer trace.Epoch("admin path")()
 		reply, err = r.addAdminCmd(ctx, args)
 	} else if proto.IsReadOnly(args) {
 		defer trace.Epoch("read-only path")()
-		reply, err = r.addReadOnlyCmd(ctx, args)
+		reply, err = r.addReadOnlyCmd(ctx, bArgs)
 	} else if proto.IsWrite(args) {
 		defer trace.Epoch("read-write path")()
-		reply, err = r.addWriteCmd(ctx, args, nil)
+		reply, err = r.addWriteCmd(ctx, bArgs, nil)
+	} else if bArgs != nil && len(bArgs.Requests) == 0 {
+		// empty batch; do nothing.
```

Comment:
not with a small fix to the `resolveWriteIntentError` call path I just added. Changing this to a `panic` (don't want this happening with #2141 anyway).